pkg/eventprocessor/impl/executor/impl/blockscope.go

```diff
@@ -162,6 +165,54 @@ func (bs *blockScope) TxnReceiptExists(ctx context.Context, txnHash common.Hash)
 	return true, nil
 }
 
+func (bs *blockScope) StateHash(ctx context.Context, chainID tableland.ChainID) (executor.StateHash, error) {
+	hash, err := dbhash.DatabaseStateHash(ctx, bs.txn, []dbhash.Option{
+		dbhash.WithFetchSchemasQuery(
+			fmt.Sprintf(`SELECT tbl_name, sql 
+				FROM sqlite_schema
+			    WHERE name NOT LIKE 'sqlite_%%'  
+				AND name LIKE '%%\_%d\_%%' ESCAPE '\'
+				AND type = 'table'
+				UNION ALL
+				SELECT tbl_name, sql 
+				FROM sqlite_schema
+				WHERE name in ('registry', 'system_acl', 'system_controller', 'system_txn_receipts')
+				ORDER BY tbl_name COLLATE nocase;`, chainID),
+		),
+		dbhash.WithPerTableQueryFn(func(tableName string) string {
+			switch tableName {
+			case "registry":
+				return fmt.Sprintf(`SELECT id, chain_id, controller, prefix, structure 
+							FROM registry 
+							WHERE chain_id = %d 
+							ORDER BY id`, chainID)
+			case "system_acl":
+				return fmt.Sprintf(`SELECT chain_id, table_id, controller, privileges 
+							FROM system_acl 
+							WHERE chain_id = %d 
+							ORDER BY table_id`, chainID)
+			case "system_controller":
+				return fmt.Sprintf(`SELECT chain_id, table_id, controller 
+							FROM system_controller 
+							WHERE chain_id = %d
+							ORDER BY table_id`, chainID)
+			case "system_txn_receipts":
+				return fmt.Sprintf(`SELECT chain_id, block_number, index_in_block, txn_hash, error, table_id 
+							FROM system_txn_receipts 
+							WHERE chain_id = %d 
+							ORDER BY table_id, block_number, index_in_block`, chainID)
+			default:
+				return fmt.Sprintf("SELECT * FROM %s", tableName)
```

Comment:
I guess this is deterministic for all user tables. (maybe adding an `ORDER BY rowid` to be sure.)